Add documentation to automate sourcemaps upload (#242)
DataDog/dd-sdk-reactnative: docs/crash_reporting.md

npm install --save-dev @datadog/datadog-ci
```

#### Automatically on each release build (React Native >= 0.69)

> Manually uploading your sourcemaps on every release build takes time and is prone to errors. We recommend automatically sending your sourcemaps every time you run a release build.

Reviewer: ```suggestion
Manually uploading your source maps on every release build takes time and is prone to errors. Datadog recommends automatically sending your source maps every time you run a release build.
```

> Create a new script file named `datadog-sourcemaps.sh` at the root of your project, containing this:

Reviewer: ```suggestion
Create a script file named `datadog-sourcemaps.sh` at the root of your project containing the following:
```

```shell

> #!/bin/sh

Reviewer 2: should we add `set -e` here the same way it appears in the script for pre-0.69 versions?

Author: ```suggestion
#!/bin/sh
set -e
```

# If the build runs from XCode, we cannot use yarn.

> # Therefore we need to check first which yarn command is appropriate

Reviewer: ```suggestion
# Check first which yarn command is appropriate
```

Author: ```suggestion
# Check first which yarn executable is appropriate 
```

package_manager_test_command="bin" # both `yarn bin` and `npm bin` are valid commands
test_and_set_package_manager_bin()
{
  $(echo $1 $package_manager_test_command) && export PACKAGE_MANAGER_BIN=$1
}

test_and_set_package_manager_bin "yarn" || # Replace yarn by npm if you use npm
test_and_set_package_manager_bin "/opt/homebrew/bin/node /opt/homebrew/bin/yarn" || # Replace yarn by npm if you use npm
echo "package manager not found"

REACT_NATIVE_XCODE="node_modules/react-native/scripts/react-native-xcode.sh"
DATADOG_XCODE="$(echo $PACKAGE_MANAGER_BIN) datadog-ci react-native xcode"

/bin/sh -c "$DATADOG_XCODE $REACT_NATIVE_XCODE"
```

This script finds the best way to run the `yarn datadog-ci react-native xcode` command:

> -   `yarn` can be used if you use a tool like [fastlane][9] or a service like [Bitrise][10] or [AppCenter][11] to build your app
> -   `/opt/homebrew/bin/node /opt/homebrew/bin/yarn` must be used on Mac if you run the release build from XCode directly

Reviewer: ```suggestion
- `yarn` can be used if you use a tool like [fastlane][9] or a service like [Bitrise][10] or [AppCenter][11] to build your app
- `/opt/homebrew/bin/node /opt/homebrew/bin/yarn` must be used on Mac if you run the release build from XCode directly
```

> It then runs this command that will take care of uploading the sourcemaps with all the right parameters for you. See [datadog-ci documentation][12] for more information.

Reviewer: ```suggestion
It runs this command that takes care of uploading the source maps with all the correct parameters. For more information, see the [datadog-ci documentation][12].
```

> Open your `.xcworkspace` with XCode, then select your project > Build Phases > Bundle React Native code and images. Edit the script to look like this:

Reviewer: ```suggestion
Open your `.xcworkspace` with XCode, then select your project > Build Phases > Bundle React Native code and images. Edit the script to look like the following:
```

```shell
set -e
WITH_ENVIRONMENT="../node_modules/react-native/scripts/xcode/with-environment.sh"
# Add these two lines
REACT_NATIVE_XCODE="./datadog-sourcemaps.sh"
export SOURCEMAP_FILE=./main.jsbundle.map

# Edit the next line
/bin/sh -c "$WITH_ENVIRONMENT $REACT_NATIVE_XCODE"
```

For the upload to work, you need to provide your Datadog API key. If you use a command-line tool or an external service, you can specify it as a `DATADOG_API_KEY` environment variable. If you run the build from XCode, create a `datadog-ci.json` file at the root of your project containing the API key:

```json
{
    "apiKey": "<YOUR_DATADOG_API_KEY>"
}
```

> You can also specify the Datadog site (e.g. `datadoghq.eu`) as a `DATADOG_SITE` environment variable, or as a `datadogSite` key in your `datadog-ci.json` file.

Reviewer: ```suggestion
You can also specify the Datadog site (such as `datadoghq.eu`) as a `DATADOG_SITE` environment variable, or as a `datadogSite` key in your `datadog-ci.json` file.
```

#### Automatically on each release build (React Native < 0.69)

> Open your `.xcworkspace` with XCode, then select your project > Build Phases > Bundle React Native code and images. Edit the script to look like this:

Reviewer: ```suggestion
Open your `.xcworkspace` with XCode, then select your project > Build Phases > Bundle React Native code and images. Edit the script to look like the following:
```

```shell
set -e

export NODE_BINARY=node
# If the build runs from XCode, we cannot use ${this.packageManager}.
# Therefore we need to check first which ${this.packageManager} command is appropriate
package_manager_test_command="bin" # both `yarn bin` and `npm bin` are valid commands
test_and_set_package_manager_bin()
{
  $(echo $1 $package_manager_test_command) && export PACKAGE_MANAGER_BIN=$1
}

test_and_set_package_manager_bin "yarn" || # Replace yarn by npm if you use npm
test_and_set_package_manager_bin "/opt/homebrew/bin/node /opt/homebrew/bin/yarn" || # Replace yarn by npm if you use npm
echo "package manager not found"

export SOURCEMAP_FILE=./build/main.jsbundle.map
$(echo $PACKAGE_MANAGER_BIN datadog-ci react-native xcode)
```

This script finds the best way to run the `yarn datadog-ci react-native xcode` command:

> -   `yarn` can be used if you use a tool like [fastlane][9] or a service like [Bitrise][10] or [AppCenter][11] to build your app
> -   `/opt/homebrew/bin/node /opt/homebrew/bin/yarn` must be used on Mac if you run the release build from XCode directly
> 
> It then runs this command that will take care of uploading the sourcemaps with all the right parameters for you. See [datadog-ci documentation][12] for more information.

Reviewer: ```suggestion
- `yarn` can be used if you use a tool like [fastlane][9] or a service like [Bitrise][10] or [AppCenter][11] to build your app
- `/opt/homebrew/bin/node /opt/homebrew/bin/yarn` must be used on Mac if you run the release build from XCode directly

It runs this command that takes care of uploading the source maps with all the correct parameters. For more information, see the [datadog-ci documentation][12].
```

For the upload to work, you need to provide your Datadog API key. If you use a command-line tool or an external service, you can specify it as a `DATADOG_API_KEY` environment variable. If you run the build from XCode, create a `datadog-ci.json` file at the root of your project containing the API key:

```json
{
    "apiKey": "<YOUR_DATADOG_API_KEY>"
}
```

> You can also specify the Datadog site (e.g. `datadoghq.eu`) as a `DATADOG_SITE` environment variable, or as a `datadogSite` key in your `datadog-ci.json` file.

Reviewer: ```suggestion
You can also specify the Datadog site (such as `datadoghq.eu`) as a `DATADOG_SITE` environment variable, or as a `datadogSite` key in your `datadog-ci.json` file.
```

### Upload JavaScript source maps on Android builds

#### Automatically on each release build

> In your `android/app/build.gradle` file, add right after the `apply from: "../../node_modules/react-native/react.gradle"` line:

Reviewer: ```suggestion
In your `android/app/build.gradle` file, add the following after the `apply from: "../../node_modules/react-native/react.gradle"` line:
```

```groovy
apply from: "../../node_modules/@datadog/mobile-react-native/datadog-sourcemaps.gradle"
```

For the upload to work, you need to provide your Datadog API key. You can specify it as a `DATADOG_API_KEY` environment variable, or create a `datadog-ci.json` file at the root of your project containing the API key:

```json
{
    "apiKey": "<YOUR_DATADOG_API_KEY>"
}
```

> You can also specify the Datadog site (e.g. `datadoghq.eu`) as a `DATADOG_SITE` environment variable, or as a `datadogSite` key in your `datadog-ci.json` file.

Reviewer: ```suggestion
You can also specify the Datadog site (such as `datadoghq.eu`) as a `DATADOG_SITE` environment variable, or as a `datadogSite` key in your `datadog-ci.json` file.
```
